Quest-3/Part-1/main.py

```diff
@@ -0,0 +1,69 @@
+def main(fileName):
+    lines = readFile(fileName)
+
+    schemEngine = [list(line.replace("\n", "")) for line in lines]
+    symbols = set(char for line in schemEngine for char in line if char != '.' and not char.isdigit())
+
+    allNumbers = findNumbers(schemEngine, symbols)
+    trueNumbers = findSymbolsAroundNumbers(schemEngine, allNumbers, symbols)
+
+    total = sum(trueNumbers)
+    print(total)
+
+def readFile(name):
+    with open(file=f"Quest-3/{name}", mode="r", encoding="utf-8") as f:
+        return f.readlines()
+
+def findNumbers(matrix, symbols):
+    visited = set()
+    numbers = {}
+
+    def exploreNumber(row, col):
+        if row < 0 or col < 0 or row >= len(matrix) or col >= len(matrix[0]) or (row, col) in visited or matrix[row][col] == '.' or matrix[row][col] in symbols:
+            return []
+
+        visited.add((row, col))
+        number = [(row, col)]
+        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
+
+        for dr, dc in directions:
+            newRow, newCol = row + dr, col + dc
+            number += exploreNumber(newRow, newCol)
+
+        return number
+
+    for i in range(len(matrix)):
+        for j in range(len(matrix[0])):
+            if matrix[i][j].isdigit() and (i, j) not in visited:
+                numberCoordinates = exploreNumber(i, j)
+                numKey = ''.join(matrix[i][j] for i, j in numberCoordinates)
+                numbers[numKey] = numberCoordinates
+
+    return numbers
+
+def findSymbolsAroundNumbers(matrix, numbers, symbols):
```

Comment:
too complex. You know the length of each number. Numbers are written horizontally, always. So you can actually compute all the coords of symbols around the number, instead of scanning around each digit.